src/test/java/com/atlassian/bitbucket/jenkins/internal/fixture/JenkinsProjectRule.java

```diff
@@ -0,0 +1,173 @@
+
+/**
+ * A test rule designed to be used as part of a test chain to provide your tests with the ability to run against multiple
+ * kinds of supported Jenkins projects, and ensure your features have compatibility without having to write a separate
+ * test case for each one. The rule is designed for black-box testing (treating any kind of Jenkins project as a generic
+ * {@link hudson.model.Job}), and can be customized to run with any chosen set of projects if your feature is not
+ * intended to have full compatibility. This rule can also be used to handle project instantiation and cleanup for individual jobs.
+ */
+public class JenkinsProjectRule implements TestRule {
+
+    // The set of all project types that can be tested by this rule, specified in the builder
+    private final Set<ProjectType> testedProjectTypes;
+    // The chained JenkinsRule, used to instantiate projects
+    private final JenkinsRule jenkinsRule;
+
+    // The job available to the integration test
+    private Job activeJob;
+    private ProjectType activeJobType;
+    // The list of all jobs that have been created as part of testing this job
+    private List<Job> instantiatedJobs = new ArrayList<>();
+
+    private JenkinsProjectRule(Builder builder) {
+        this.jenkinsRule = builder.jenkinsRule;
+        this.testedProjectTypes = builder.projectTypes;
+    }
+
+    @Override
+    public Statement apply(Statement base, Description description) {
+        return new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                // If the method was not annotated with a project initialization, run as normal
+                if (testedProjectTypes.isEmpty()) {
+                    base.evaluate();
+                } else {
+                    try {
+                        initializeProjects();
+                        for (ProjectType projectType : testedProjectTypes) {
+                            setActiveJob(projectType);
+                            base.evaluate();
+                        }
+                    } finally {
+                        cleanupProjects();
+                    }
+                }
+            }
+        };
+    }
+
+    /**
+     * Returns the job that the project rule is currently testing.
+     *
+     * @return the currently active job
+     */
+    public Job getJob() {
+        return activeJob;
+    }
+
+    /**
+     * Sets the SCM of the currently active job
+     *
+     * @param scm the SCM to add to the currently active job
+     */
+    public void setSCM(SCM scm) throws IOException {
+        switch(activeJobType) {
+            case FREESTYLE:
+                ((FreeStyleProject) activeJob).setScm(scm);
+                break;
+            case PIPELINE:
+                ((WorkflowJob) activeJob).setDefinition(new CpsScmFlowDefinition(scm, "Jenkinsfile"));
+                break;
+            case MULTIBRANCH:
+                // TODO: Add Multibranch Support
```

Comment:
No Mutlibranch yet. This will be addressed in a separate PR later. :disappointed: